### PR TITLE
I40 component ingredent params bug

### LINKED ISSUE #27: [BUG][Templates] Components should use the user defined element and not create a new one
**Describe the bug**
Right now, if someone includes a component as an element, it'll create it as a _sibling_ to that element with the component name as the tag name instead of as that element the user defined. (See screenshot) 

It's extra bloat on the dom and it's a little confusing that if the name of the component is something like "inventory", the generated component is "inventorycomponent".

**Expected behavior**
The created element is the same name as the user defined element and not a sibling.

**Screenshots**
image

### LINKED ISSUE #28: [BUG][Component] Components with multiple words cannot fetch template and styling
**Describe the bug**
When you have a component with more than two words + "Component" as the name, it cannot properly fetch the template and stylesheet as it's matching against the wrong string.

There's a regEx string that converts strings like MyReallyCoolComponent into my-really-cool, but it's only matching against the first instance, but not the proceeding ones.

The issue lies here:
Component.class.ts:
image

The dirname is incorrect because the regEx match is not greedy and only does the replacement on the first match.

**To Reproduce**
1. Create a component like "MyReallyCoolComponent". Add a template or style file along with it.
2. Reference this component in another component. It will not show up.

**Expected behavior**
MyReallyCoolComponent should look for my-really-cool.component.html and my-really-cool.component.(css/scss)

### LINKED ISSUE #30: [BUG][Template] Components with multiple words does not load from templates
**Describe the bug**

**To Reproduce**
- Have a component with multiple words like "MyCoolComponent"
- Load this component from the parent template like this:
```<my-cool></my-cool>```
- Component is not loaded.

**Expected behavior**
Dashes should be allowed in component name. Current work around is to reference the component with one lowercase word.

**Additional context**
The issue likely lives in the component-factory.class.ts file under the createRecipe fn:
```javscript
    createRecipe(parentNode) {
        if (!this.recipeCreated) {
            // Create recipe
            // Get registry and parse for any elements with custom tags names
            const moduleFactory = window.vivi;
            moduleFactory.getComponentRegistry().forEach(reg => {
                // Strip 'Component' off of name
                const name = reg.slice(0, reg.lastIndexOf('Component'));
                const els = parentNode.querySelectorAll(name.toLowerCase());
```
The "const name" section will likely need to match against a regEx replace like in issue #28. Consider creating some shared function so naming between the two are consistent.

### LINKED ISSUE #34: [FEATURE][Component] Easily add children components
**Is your feature request related to a problem? Please describe.**
Currently to append child it's a bit wordy:
```js
const editFactory: ViviComponentFactory<EditableComponent> = (<any>window).vivi.getFactory(EditableComponent);
editFactory.create({ id: this.data.shipmentId, value: this.data.weight, hasButton: true, label: 'Weight' }).append(this.weightSpan);
```

**Describe the solution you'd like**
Something similar to events where they can be added and destroyed along with the parent would be handy.

**Additional context**
Bug #26 is also relevant as the service it short hand for window.vivi.getFactory.

### LINKED ISSUE #40: [BUG][Component] Component Ingredients do not change parameters when instancing multiple
**To Reproduce**
1. Create multiple child elements in the template
2. Give both of them different parameters
3. Only the first set of parameters are used

**Expected behavior**
Data assigned is assigned to the correct component

### PR DESCRIPTION
**Pull request by CLeeSpruit**
### Linked Issues:
Fixes #40 
Closes #34

#### Meta
Created a "Mocker" that mocks out the module and generates custom components for tests. This eliminates all the "MockComponent" variations around the repo.

#### Component Factory
- Style is now added in the constructor, rather than the first component creation
- Component recipe (child components from the template) parsing moved to Parse Engine
- Create function greatly simplified and most of the logic moved to the component
- Destroy will throw an error if no component is found with given id
- Destroy checks if the component is in the DOM before removing
- Added DestroyAll fn to bulk remove all components from a factory (used mostly in the component creator)

#### Parse Engine
- Component children from template are now determined on Parse
- Parse Engine changed to a service and split into elements & components
- Node typing moved to HTMLElement for more consistency and accuracy

#### Component
- Component name added to consistently get element name (my-cool)
- Node types (ogNode, parsedNode) moved to use HTMLElement, which is what it is
- Removed isLoaded and isVisible
    - *Breaking Change*: Nothing internally was dependent on this, but if an external application used this, you'll have to use other means.
- Recipe renamed to "children" to include dynamically created components
- Parent set as document.body by default
- Create Recipe Fn removed (moved to parse engine)
ogNode's innerHTML and trigger a redraw.
- Merged beforeLoadHook into loadAll
- Destroy now also destroys children
- Added CreateChild fn to easily add components of any type (primarily fixes #34)

#### Component Params
- Params class was removed and just changed into an object

#### Component Ingredient
- Ingredient class was removed. All other instances were changed to child components.

### Type:
This is a minor update. While there's a lot of refactoring, most of the end changes are very little.

### CODE REVIEW

File: src/decorators/element.decorator.ts [hunk 1]
```diff
@@ -10,17 +10,17 @@ interface ViviElementParams {
...
             props = [objParams];
-            Reflect.defineMetadata(metadataKey, props, target);
+            Reflect.defineMetadata(elementMetadataKey, props, target);
         }
     }
 }
 
 export function getElements(origin: Object): Array<ViviElementParams> {
-    return Reflect.getMetadata(metadataKey, origin) || new Array<ViviElementParams>();
+    return Reflect.getMetadata(elementMetadataKey, origin) || new Array<ViviElementParams>();
```
Comment (CLeeSpruit) [inline src/decorators/element.decorator.ts:25]: Most of the changes to this file is for the component creator. May possibly clean up.

File: src/factory/__spec__/component-factory.class.spec.ts [hunk 2]
```diff
@@ -1,135 +1,90 @@
...
-            mock.destroy(comp.id);
+            creator.getFactory().destroy(comp.id);
 
-            const actual = mock.get(comp.id);
+            const actual = creator.getFactory().get(comp.id);
             expect(actual).toBeFalsy();
         });
     });
 });
-
```
Comment (CLeeSpruit) [inline src/factory/__spec__/component-factory.class.spec.ts:115]: This all got replaced by Component Creator.

File: src/models/__mocks__/component-creator.class.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    hasElements?: boolean;
+    elements?: Array<ViviElementParams>;
+}
+
+export class ComponentCreator {
+    module: ModuleFactory;
+    readonly defaultComponents = [
+        { constructor: MockComponent }
+    ];
+
```
Comment (CLeeSpruit) [inline src/models/__mocks__/component-creator.class.ts]: Clean up extra space

File: src/models/__spec__/component-params.spec.ts [hunk 4]
```diff
@@ -1,35 +1,33 @@
-import { ModuleFactory } from '../../factory';
```
Comment (CLeeSpruit) [inline src/models/__spec__/component-params.spec.ts:1]: I think it may be worth looking at turning ComponentParams into an interface (or nothing?). The constructor is rarely used, if at all anymore.

File: src/models/component-ingredient.class.ts [hunk 5]
```diff
@@ -3,23 +3,18 @@ import { ViviComponentFactory } from 'factory';
...
-        this.component = this.factory.create(this.params);
-        this.component.append(this.element.parentElement, true);
-    }
-
-    load(parent: Node) {
-        this.component.loadAll(parent);
+    load() {
+        this.component.loadAll();
     }
 }
```
Comment (CLeeSpruit) [inline src/models/component-ingredient.class.ts]: I wonder if the ingredients are even need at all any more? That would simplify some code if it could just straight go to creating components.

File: src/models/component.class.ts [hunk 6]
```diff
@@ -90,64 +95,58 @@ export abstract class Component {
...
-        if (this.element) {
-            const newNode = ParseEngine.parseNode(this.ogNode, this.data);
-            const oldNode = document.getElementById(this.id);
-            this.parent.replaceChild(newNode, oldNode);
-            this.parsedNode = newNode;
-            this.element = document.getElementById(this.id);
+        // Remove 
+        const newParse = ParseEngine.parseNode(this.ogNode, this.data);
+        const newEl = newParse.el;
+        const recipe = newParse.recipe;
```
Comment (CLeeSpruit) [inline src/models/component.class.ts]: Recipe is just left hanging. This needs to somehow replace the previous children.

File: src/models/__spec__/component-params.spec.ts (hunk 4 above)
Comment (CLeeSpruit) [reply src/models/__spec__/component-params.spec.ts:1]: Removed and swapped with just an object

File: src/models/component-ingredient.class.ts (hunk 5 above)
Comment (CLeeSpruit) [reply src/models/component-ingredient.class.ts]: Actually yeah. This cleans up a ton of unneeded complexity.
Done.

File: src/models/__mocks__/component-creator.class.spec.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        });
+
+        it('default - should return MockComponent', () => {
+            const mock = creator.createMock();
+
+            expect(mock).toBeTruthy();
+            expect(mock).toBeInstanceOf(MockComponent);
+        });
+
+        it('hasTemplate: true - should return component with default template', () => {
```
Comment (CLeeSpruit) [inline src/models/__mocks__/component-creator.class.spec.ts]: It might be worth automating this section.

Comment (CLeeSpruit) [reply src/models/__mocks__/component-creator.class.spec.ts]: Spoiler: It was not.

File: src/models/component.class.ts (hunk 6 above)
Comment (CLeeSpruit) [reply src/models/component.class.ts]: This is being left as it's own issue rather than dragging it into this one.

**Comment by CLeeSpruit** [comment]
tl;dr - Can't have a bug if the code doesn't exist the first place.

Tap Head Meme
